Resource name generation

maltesander opened the pull request.

razvan commented on src/name_utils.rs:
```diff
@@ -0,0 +1,253 @@
+
+    let min = kind.min();
+    if alphanumeric.len() < usize::from(min) {
+        return Err(Error::SubNameTooShort {
+            kind: kind.to_string(),
+            name: sub_name.to_string(),
+            min,
+        });
```
How about padding the sub_name with nulls (0) for example instead of erroring because it's too small ? It's what the caller would probably have to do anyway.

razvan commented on src/name_utils.rs:
```diff
@@ -0,0 +1,253 @@
+    node_name: Option<&str>,
+    misc_name: Option<&str>,
+) -> OperatorResult<String> {
+    let mut full_name = String::new();
```
If `full_name` were a vec you could just append to it and call `join("-")` at the end.

razvan commented on src/name_utils.rs:
```diff
@@ -0,0 +1,253 @@
+    Ok(alphanumeric
+        .chars()
+        .into_iter()
+        .take(usize::from(kind.max()))
```
If characters are dropped, a  warning should probably be logged out.

maltesander commented:
I changed the process a little. Fixed length (min, max) are removed. Now each sub name receives an equal share of the possible characters (63 - reserved_for_hash - len(sub_names) [which are the dashes we add later]). If a sub_name is shorter than its share the characters may be used for another sub name.

razvan commented on src/name_utils.rs:
```diff
@@ -0,0 +1,399 @@
+///                 `RESOURCE_NAME_MAX_LEN` - `KUBERNETES_HASH_MIN_LENGTH` - sub_names.len()
+///
+fn selectable_chars(sub_names: &[String], max_chars: usize) -> usize {
+    (1f32 / sub_names.len() as f32 * max_chars as f32).floor() as usize
```
This is losing remainder characters. I think this is better:

```suggestion
        max_chars / sub_names.len() + max_chars % sub_names.len()
```

After this change, two tests need to be updated.

maltesander replied:
Done.

razvan commented on src/name_utils.rs:
```diff
@@ -0,0 +1,399 @@
+        }
+
+        full_name.push_str(&selected);
+        full_name.push('-');
```
This always adds a '-' at the end. Not  good. Use a `Vec<String>` for `full_name` and do `full_name.join("-")` at the end.

maltesander replied:
Yes thats what we want. It is intended to work with generatedName from kubernetes, which will append a hash at the end. Wont look good without that.

razvan commented on src/name_utils.rs:
```diff
@@ -0,0 +1,399 @@
+const RESOURCE_NAME_MAX_LEN: usize = 63;
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
```
Just wondering why these are not in error.rs.

maltesander replied:
I like to have fewer operator errors and derive from smaller errors. Its just a preference, no real difference. Since its not an external library it would make sense in the main errors though.

razvan commented on src/name_utils.rs:
```diff
@@ -0,0 +1,399 @@
+
+    // Calculate the maximum number of available characters
+    // The sub names vector length equals the size of additional required dashes
+    let max_chars = RESOURCE_NAME_MAX_LEN - KUBERNETES_HASH_MIN_LENGTH - sub_names.len();
```
You can add one (+1) to this if you remove the trailing '-' that is always added now.

maltesander replied:
It will stay (check above).

razvan reviewed (approved):
I didn't realize that we are not building the final resource name. Now I understand. It looks good.
